### PR TITLE
soc/intel_adsp: support 2 and 4 DSP core cavs25 variants

### PR DESCRIPTION
**Pull request by kv2019i**
New attempt to unblock our validation on 4 DSP core systems. Previous attempt at https://github.com/zephyrproject-rtos/zephyr/pull/40960 .

I note the core count is taken from Kconfig in all cAVS platforms currently, so I did not proceed to change this model (and this would need to be a separate PR). Instead I added a new board for the 2core variant, fixed a few issues in common code where it is assumed all 4 cores are always defined in dts, and last, enabled all 4 cores for original cavs25 target.

Is this in the right ballpark @andyross @MaureenHelm .

### CODE REVIEW

File: soc/xtensa/intel_adsp/common/irq-cavs.c
Comment (kv2019i) [inline soc/xtensa/intel_adsp/common/irq-cavs.c:35]: Please suggest if there a better way to make conditional definitions.

Comment (andyross) [reply soc/xtensa/intel_adsp/common/irq-cavs.c:35]: Honestly?  Refactor the interrupt controller device so that it actually manages all the interrupts on the system and doesn't have to be instantiated once per core.   Where I've had the opportunity, I've done exactly that. :)

This is historical: early cavs versions in Zephyr were single core and managed a single register block, so the "obvious" (but, well, wrong) way to implement SMP was to just make more of them.  And so we pay in future complexity.

For now this is probably the best that can be done in a focused patch.  Though if you want to be clever there are macros defined that can enumerate over defined DT instances.  But rework and refactoring is definitely the current zeitgeist on these platforms, so...

File: dts/xtensa/intel/intel_cavs25_2core.dtsi
Comment (andyross) [inline dts/xtensa/intel/intel_cavs25_2core.dtsi:20]: Not that it really matters, but FWIW: nothing in Zephyr actually uses these cpu definitions.  We still get core count out of kconfig, and to the extent we'd want to put it in devicetree, we'd want a cavs-specific node here that tells us how many cores exist.  Having multiple cpu devices only makes sense if they can have divergent configuration (e.g. the ULP core? maybe that would make sense).

File: dts/xtensa/intel/intel_cavs25_2core.dtsi
Comment (andyross) [inline dts/xtensa/intel/intel_cavs25_2core.dtsi:33]: That said... if it's a 2-core device you should probably remove cpu2 and cpu3 :)

File: soc/xtensa/intel_adsp/common/irq-cavs.c
Comment (andyross) [reply soc/xtensa/intel_adsp/common/irq-cavs.c:35]: Oh, heh, except that I forget: this is cavs.  :)  The multitple definitions aren't per-core, they're per-level.  So... they should still be present in this device, no?  The "cavs2" devicetree node defines a controller that listens to the third of four external interrupts on the device.

File: boards/xtensa/intel_adsp_cavs25_2core/Kconfig.defconfig
Comment (andyross) [inline boards/xtensa/intel_adsp_cavs25_2core/Kconfig.defconfig:17]: Don't think these values change, do they?  Probably best to leave them defined at the SOC level somehow than cut/paste.

**Review by andyross — COMMENTED** [review]
I'm a little worried about the interrupt controller changes, though maybe I'm misunderstanding the hardware.

File: boards/xtensa/intel_adsp_cavs25_2core/Kconfig.defconfig
Comment (kv2019i) [reply boards/xtensa/intel_adsp_cavs25_2core/Kconfig.defconfig:17]: Let me check. I'll at least refer to the base cavs25 definitions. Agreed these should not be copy'n'pasted here.

File: dts/xtensa/intel/intel_cavs25_2core.dtsi
Comment (kv2019i) [reply dts/xtensa/intel/intel_cavs25_2core.dtsi:20]: I'll keep core count also in the kconfig option for now in this PR

File: dts/xtensa/intel/intel_cavs25_2core.dtsi
Comment (kv2019i) [reply dts/xtensa/intel/intel_cavs25_2core.dtsi:33]: Yes, fixed.

**Comment by kv2019i** [comment]
Sorry for the long delay in updating the series. Now refreshed based on feedback and review for another round of review.

I dropped the patch to bump the core count, so this can be merged without interfering with existing users. I will send that later when I can move e.g. SOF to use the new 2core variant and only use the original config for targets that really can handle 4 cores.

Series tested with both 2-core and 4-core variants for cAVS2.5.

File: soc/xtensa/intel_adsp/common/irq-cavs.c
Comment (MaureenHelm) [reply soc/xtensa/intel_adsp/common/irq-cavs.c]: > Though if you want to be clever there are macros defined that can enumerate over defined DT instances. 

I think you could use these macros to generate a table and remove the switch block entirely. Have a look at `DT_INST_FOREACH_STATUS_OKAY`

Comment (kv2019i) [reply soc/xtensa/intel_adsp/common/irq-cavs.c]: @andyross @MaureenHelm I dived into the spec a bit today and indeed the controller bits are always there, independently of the CPU cores, so it would seem easiest solution (and matching hw) is to just keep the the definitions although for the less-CPU-cores variant. This also avoids dropping all changes to this irq-cavs.c file. This is now reflected in the latest revision.

**Comment by kv2019i** [comment]
V3 uploaded. Only change is the name of the new config. Now using Intel PCH name to identify the board.

**Review by andyross — DISMISSED** [review]
Makes sense to me.  As mentioned, I do think that we need to do some rework still to reduce the amount of cut/paste needed when redefining boards like this[1].  The devicetree layer in particular requires way too much code just to change core counts (which aren't even stored in devicetree yet!).

[1] Though if I can toot my own horn a bit: we no longer have to copy thousands of lines of header files and the C code is fully unified now.  But for the most part I only did cosmetic level cleanup of the configuration layer.

**Review by dcpleung — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

File: samples/subsys/audio/sof/boards/intel_adsp_cavs25.conf
Comment (MaureenHelm) [inline samples/subsys/audio/sof/boards/intel_adsp_cavs25.conf:7]: Will you please submit this commit in a separate PR so it can be merged before the 3.0 release?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
What's there looks good, but a board doc needs to be added.

File: samples/subsys/audio/sof/boards/intel_adsp_cavs25.conf
Comment (kv2019i) [reply samples/subsys/audio/sof/boards/intel_adsp_cavs25.conf:7]: @MaureenHelm This is harmless now and we can wait until the full thing is merged post-3.0. Let me work on the board doc. Basicly we currently really only support the 2core variant (tgl-h) in Zephyr and you can run a tgl-h binary on tgl hardware, so this ok.

**Comment by kv2019i** [comment]
Board documentation file now added.

**Review by andyross — DISMISSED** [review]
I can't see why not to merge this.  It's absolutely doing what all the existing variants do.  It's true that there's lots of cut and paste here, but none of that is the fault of the PCH variant, or SOF.  Let's merge it and clean it up once it's in-tree.

**Review by dcpleung — DISMISSED** [review]

**Comment by kv2019i** [comment]
Hmm, one build error:
```
Warning, treated as error:
/home/runner/work/zephyr/zephyr/doc/_build/src/boards/xtensa/intel_adsp_cavs25_tgph/doc/index.rst:14:Unknown target name: "zephyr audio dsp development on chromebooks".
```

Let me update.

**Review by teburd — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Comment by kv2019i** [comment]
@kv2019i  wrote:
> Hmm, one build error:

Now fixed, all CI tests pass.

File: boards/xtensa/intel_adsp_cavs25_tgph/board.cmake
Comment (marc-hb) [inline boards/xtensa/intel_adsp_cavs25_tgph/board.cmake:6]:
```suggestion
board_set_rimage_target(tgl-h)
```

Comment (kv2019i) [inline boards/xtensa/intel_adsp_cavs25_tgph/board.cmake:6]: Thanks @marc-hb , will fix.

**Comment by kv2019i** [comment]
One more update, fixed s/tgl/tgl-h/ in board.cmake. No other changes.

File: boards/xtensa/intel_adsp_cavs25_tgph/intel_adsp_cavs25_tgph.dts
Comment (mbolivar-nordic) [inline boards/xtensa/intel_adsp_cavs25_tgph/intel_adsp_cavs25_tgph.dts:13]: I don't see much point to this compatible, but it's not worth blocking over.

Comment (kv2019i) [reply boards/xtensa/intel_adsp_cavs25_tgph/intel_adsp_cavs25_tgph.dts:13]: I kept this still as the same compatible is in all cavs variants. better to remove from all if we remove.

**Comment by kv2019i** [comment]
New version uploaded to rebase to v3.0.0 changes. Please review.

**Comment by kv2019i** [comment]
@andyross @MaureenHelm  @nashif  It seems this PR has got stuck, Github is not noticing updates to my branch. Should I open a new PR for this, or what's the BKM in these cases in Zephyr.

**Comment by kv2019i** [comment]
Queued a draft-PR https://github.com/zephyrproject-rtos/zephyr/pull/43104 while waiting for guidance.

**Comment by marc-hb** [comment]
> It seems this PR has got stuck, Github is not noticing updates to my branch.

There's a git conflict now, maybe Github didn't notice yet. I'm curious how often Github refreshes pull/41764/merge.

**Pull request closed without merging**
